openupgrade_scripts/scripts/sale_loyalty/16.0.1.0/pre-migration.py

```diff
@@ -0,0 +1,299 @@
+        "sale_loyalty.sale_order_view_form",
+        "sale_loyalty.sale_order_view_form_inherit_sale_loyalty",
+    ),
+]
+
+
+def _generate_random_reward_code():
+    """This function, _generate_random_reward_code, provides a random reward code. It is
+    used to create unique codes that identify rewards within the sale_loyalty module."""
+    return str(random.getrandbits(32))
+
+
+def convert_applied_coupons_from_sale_order_to_many2many(env):
+    """This conversion from One-to-Many (o2m) to Many-to-Many (m2m) involves the
+    creation of a relationship table between sale_order and loyalty_card, allowing
+    multiple sale_order records to be associated with multiple loyalty_card records.
+    This is done by inserting the ids corresponding to sale_order and loyalty_card
+    into the loyalty_card_sale_order_rel table."""
+    openupgrade.logged_query(
+        env.cr,
+        """
+        CREATE TABLE IF NOT EXISTS loyalty_card_sale_order_rel (
+            sale_order_id INT,
+            loyalty_card_id INT
+        )
+        """,
+    )
+    openupgrade.logged_query(
+        env.cr,
+        """
+        INSERT INTO loyalty_card_sale_order_rel (sale_order_id, loyalty_card_id)
+        SELECT sales_order_id, id
+        FROM loyalty_card
+        """,
+    )
+
+
+def fill_code_enabled_rule_ids_from_sale_order(env):
+    """This is an m2m field, the relationship table is created in case it does not
+    exist and filled with the data. This field relates the id of a rule to the id of
+    a sales order to which it has been applied, provided that the rule belongs to an
+    applicable promotion with code."""
+    openupgrade.logged_query(
+        env.cr,
+        """
+        CREATE TABLE IF NOT EXISTS loyalty_rule_sale_order_rel (
+            sale_order_id INT,
+            loyalty_rule_id INT
+        )
+        """,
+    )
+    openupgrade.logged_query(
+        env.cr,
+        """
+        INSERT INTO loyalty_rule_sale_order_rel (sale_order_id, loyalty_rule_id)
+        SELECT lc.order_id, lr.id
+        FROM loyalty_card lc
+        JOIN loyalty_rule lr ON lc.program_id = lr.program_id
+        JOIN loyalty_program lp ON lr.program_id = lp.id
+        WHERE (lr.mode = 'with_code' OR lp.program_type = 'gift_card')
+        """,
+    )
+
+
+def generate_sale_order_coupon_points(env):
+    openupgrade.logged_query(
+        env.cr,
+        """
+        CREATE TABLE IF NOT EXISTS sale_order_coupon_points (
+            coupon_id INT,
+            order_id INT,
+            points FLOAT,
+            create_uid INT,
+            write_uid INT,
+            create_date DATE,
+            write_date DATE
+        )
+        """,
+    )
+    openupgrade.logged_query(
+        env.cr,
+        """
+        INSERT INTO sale_order_coupon_points (
+            coupon_id, order_id, points, create_uid, write_uid, create_date, write_date
+        )
+        SELECT
+            id AS coupon_id,
+            order_id,
+            1 AS points,
```

Comment:
Why putting 1 point per coupon?

Reply:
I had confused it with the default points of the points program, in this case the points come from loyalty_card.